File: mobile/lib/domain/entities/victim.dart
import 'location_fix.dart';
import 'sync_status.dart';
import 'triage_category.dart';
import 'victim_demographics.dart';
import 'victim_status.dart';

/// A person registered by a responder on this device.
///
/// The record is complete and authoritative the moment it is written to local
/// storage. Nothing here waits on, or is corrected by, the backend.
class Victim {
  const Victim({
    required this.id,
    required this.temporaryId,
    required this.ageGroup,
    required this.gender,
    required this.triageCategory,
    required this.priority,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    required this.createdBy,
    required this.syncStatus,
    this.name,
    this.age,
    this.medicalCondition,
    this.injuryType,
    this.assistanceRequired,
    this.incidentId,
    this.latitude,
    this.longitude,
    this.locationAccuracy,
  });

  /// Device-minted UUID. Stable for the life of the record, across devices.
  final String id;

  /// Short human-readable label, e.g. `V-8C1F-007`. What goes on the triage
  /// tag and gets read out over the radio.
  final String temporaryId;

  /// Null when the person could not be identified, which is common and is not
  /// a reason to delay registering them.
  final String? name;

  final int? age;
  final AgeGroup ageGroup;
  final Gender gender;
  final String? medicalCondition;
  final String? injuryType;
  final TriageCategory triageCategory;

  /// Mirrors [TriageCategory.priority]. Persisted so the list can be ordered
  /// by the database rather than in memory.
  final int priority;

  final String? assistanceRequired;
  final VictimStatus status;

  /// The incident the casualty was registered under. Null on records authored
  /// by Slice 2, and on any record written before the device adopted an
  /// incident — a casualty in front of you outranks bookkeeping.
  final String? incidentId;

  /// Where the responder was standing when they registered the casualty, taken
  /// from the device's own receiver. Null when no fix was available, which is
  /// never a reason to refuse the registration.
  final double? latitude;
  final double? longitude;
  final double? locationAccuracy;

  final DateTime createdAt;
  final DateTime updatedAt;

  /// Identifier of the responder session that authored the record. Carries an
  /// offline demo session's local id unchanged, so provenance survives.
  final String createdBy;

  final SyncStatus syncStatus;

  /// What to show when the person has no recorded name.
  String get displayName {
    final trimmed = name?.trim();
    return (trimmed == null || trimmed.isEmpty) ? 'Unidentified' : trimmed;
  }

  bool get hasPosition => latitude != null && longitude != null;

  /// The stored coordinates as a reading, so the detail screen can reuse the
  /// degree and accuracy formatting rather than restating it.
  ///
  /// Timestamped with [createdAt] because that is when the position was
  /// attached; the reading itself may have been taken slightly earlier.
  LocationFix? get position => hasPosition
      ? LocationFix(
          latitude: latitude!,
          longitude: longitude!,
          accuracy: locationAccuracy,
          timestamp: createdAt,
        )
      : null;

  Victim copyWith({
    String? name,
    bool clearName = false,
    int? age,
    bool clearAge = false,
    AgeGroup? ageGroup,
    Gender? gender,
    String? medicalCondition,
    bool clearMedicalCondition = false,
    String? injuryType,
    bool clearInjuryType = false,
    TriageCategory? triageCategory,
    String? assistanceRequired,
    bool clearAssistanceRequired = false,
    VictimStatus? status,
    String? incidentId,
    double? latitude,
    double? longitude,
    double? locationAccuracy,
    DateTime? updatedAt,
    SyncStatus? syncStatus,
  }) {
    final category = triageCategory ?? this.triageCategory;
    return Victim(
      id: id,
      temporaryId: temporaryId,
      name: clearName ? null : (name ?? this.name),
      age: clearAge ? null : (age ?? this.age),
      ageGroup: ageGroup ?? this.ageGroup,
      gender: gender ?? this.gender,
      medicalCondition: clearMedicalCondition
          ? null
          : (medicalCondition ?? this.medicalCondition),
      injuryType: clearInjuryType ? null : (injuryType ?? this.injuryType),
      triageCategory: category,
      priority: category.priority,
      assistanceRequired: clearAssistanceRequired
          ? null
          : (assistanceRequired ?? this.assistanceRequired),
      status: status ?? this.status,
      incidentId: incidentId ?? this.incidentId,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      locationAccuracy: locationAccuracy ?? this.locationAccuracy,
      createdAt: createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      createdBy: createdBy,
      syncStatus: syncStatus ?? this.syncStatus,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Victim &&
          other.id == id &&
          other.temporaryId == temporaryId &&
          other.name == name &&
          other.age == age &&
          other.ageGroup == ageGroup &&
          other.gender == gender &&
          other.medicalCondition == medicalCondition &&
          other.injuryType == injuryType &&
          other.triageCategory == triageCategory &&
          other.priority == priority &&
          other.assistanceRequired == assistanceRequired &&
          other.status == status &&
          other.incidentId == incidentId &&
          other.latitude == latitude &&
          other.longitude == longitude &&
          other.locationAccuracy == locationAccuracy &&
          other.updatedAt == updatedAt &&
          other.syncStatus == syncStatus;

  @override
  int get hashCode => Object.hash(
        id,
        temporaryId,
        name,
        age,
        ageGroup,
        gender,
        medicalCondition,
        injuryType,
        triageCategory,
        status,
        assistanceRequired,
        incidentId,
        latitude,
        longitude,
        updatedAt,
        syncStatus,
      );
}

File: backend/app/models/victim.py
"""A casualty registered by a responder.

The handset is the system of record while a device is offline, so this table
is a *destination* for records that already exist elsewhere: the primary key
is the UUID minted on the device, not one generated here.
"""

from __future__ import annotations

from sqlalchemy import Enum, Float, Index, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, TimestampMixin, UuidPrimaryKeyMixin
from app.models.enums import AgeGroup, Gender, TriageCategory, VictimStatus


def _enum(enum_cls: type, name: str) -> Enum:
    return Enum(
        enum_cls,
        name=name,
        values_callable=lambda cls: [member.value for member in cls],
    )


class Victim(UuidPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "victims"

    # Written on a triage tag and read out over the radio, e.g. ``V-8C1F-007``.
    temporary_id: Mapped[str] = mapped_column(String(32), unique=True, index=True, nullable=False)

    # Unidentified casualties are the common case in the first hours and are
    # never a reason to hold up registration.
    name: Mapped[str | None] = mapped_column(String(160), nullable=True)
    age: Mapped[int | None] = mapped_column(Integer, nullable=True)
    age_group: Mapped[AgeGroup] = mapped_column(
        _enum(AgeGroup, "age_group"), nullable=False, default=AgeGroup.UNKNOWN
    )
    gender: Mapped[Gender] = mapped_column(
        _enum(Gender, "gender"), nullable=False, default=Gender.UNKNOWN
    )
    medical_condition: Mapped[str | None] = mapped_column(Text, nullable=True)
    injury_type: Mapped[str | None] = mapped_column(String(200), nullable=True)
    triage_category: Mapped[TriageCategory] = mapped_column(
        _enum(TriageCategory, "triage_category"), nullable=False
    )

    # Denormalised :attr:`TriageCategory.priority` so ordering by severity is a
    # plain indexed sort rather than a CASE expression.
    priority: Mapped[int] = mapped_column(Integer, nullable=False)

    assistance_required: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[VictimStatus] = mapped_column(
        _enum(VictimStatus, "victim_status"), nullable=False, default=VictimStatus.REGISTERED
    )

    # Reserved for the spatial slice; devices send null until then.
    latitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    longitude: Mapped[float | None] = mapped_column(Float, nullable=True)

    # The authoring session, kept as free text because an offline demo session
    # has a device-local id that does not exist in ``users``.
    created_by: Mapped[str] = mapped_column(String(64), nullable=False)

    __table_args__ = (Index("ix_victims_priority_updated_at", "priority", "updated_at"),)

    def __repr__(self) -> str:  # pragma: no cover - debugging aid
        return f"<Victim {self.temporary_id} {self.triage_category.value}>"
